### ISSUE TITLE
overloaded method that has `NoReturn` for certain arguments not detected 

### ISSUE THREAD

**Issue by Dr-Irv** [python/mypy#15861]
**Bug Report**

If you have a method that has an overload for specific arguments that are not allowed, by having that overload return `NoReturn`, and a wider default that accepts `Any`, mypy does not detect the invalid argument.

**To Reproduce**

```python
from typing import Any, NoReturn, overload


class Goo:
    @overload
    def __add__(self, other: str) -> NoReturn:
        ...

    @overload
    def __add__(self, other: Any) -> bool:
        ...

    def __add__(self, other: Any) -> bool:
        return True


y = Goo()
y + 4
y + "abc"
```

**Expected Behavior**

I expect that the line `y + "abc"` is reported as an invalid operation.

**Actual Behavior**

```text
Success: no issues found in 1 source file
```

**Your Environment**

- Mypy version used:  1.4.1
- Mypy command-line flags:  None
- Mypy configuration options from `mypy.ini` (and other config files):  None
- Python version used:  3.9.16

**Comment by JelleZijlstra** [comment]
Mypy is correct here. NoReturn is not a way to make the type checker emit errors; it indicates that the code path never returns. The standard example is `sys.exit()`: it returns `NoReturn`, but type checkers should not emit an error for every use of `NoReturn`.

You are looking for python/typing#1043.

**Comment by erictraut** [comment]
I agree with Jelle's analysis. I just noticed that pyright contains a bug that causes it to (incorrectly) emit an error when a binary operator like `__add__` returns `NoReturn`. It should not emit an error here, consistent with mypy.

**Comment by Dr-Irv** [comment]
> I agree with Jelle's analysis. I just noticed that pyright contains a bug that causes it to (incorrectly) emit an error when a binary operator like `__add__` returns `NoReturn`. It should not emit an error here, consistent with mypy.

Interesting. I liked the pyright behavior in this case!  That’s why I reported the issue here.
